teps/0075-object-param-and-result-types.md

```diff
@@ -0,0 +1,924 @@
+for accessing "child" members of the object by name in variable replacement.
+
+For example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+...
+spec:
+  params:
+    - name: gitrepo
+      type: object
+      properties:
+        url: {}
+        commitish: {}
+...
+tasks:
+  - name: notify-slack-before
+    params:
+      - name: message
+        value: "about to clone $(params.gitrepo.url) at $(params.gitrepo.commitish)"
+  - name: clone-git
+    runAfter: [ 'notify-slack' ]
+    params:
+      - name: gitrepo
+        value: $(params.gitrepo[*])
+    results:
+      - name: cloned-gitrepo
+        type: object
+        properties:
+          url: {}
+          commitish: {}
+    taskSpec:
+      params:
+        - name: gitrepo
+          type: object
+          properties:
+            url: {}
+            commitish: {}
+      steps:
+        - name: do-the-clone
+          image: some-git-image
+          args:
+            - "-url=$(params.gitrepo.url)"
+  - name: notify-slack-after
+    params:
+      - name: message
+        value: "cloned $(tasks.cloned-git.results.cloned-gitrepo.url) at $(tasks.cloned-git.results.cloned-gitrepo.commitish)"
+```
+
+_This proposal does not include adding support for any additional syntax (though it could be added in the future!)._
+
+#### Variable replacement with object params
+
+**When providing values for objects** Task and Pipeline authors can provide an entire object as a value only when the
+value is also an object (see [matching interface and extra keys](#matching-interfaces-and-extra-keys)) by using
+[the same `[*]` syntax used to provide entire arrays](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#substituting-array-parameters)
+.
+
+In the above example:
+
+```yaml
+params:
+  - name: gitrepo
+    value: $(params.gitrepo[*])
+```
+
+**When providing values for strings**, Task and Pipeline authors can access individual attributes of an object param;
+they cannot access the object as whole (we could add support for this later).
+
+In the above example, within a Pipeline Task:
+
+```yaml
+value: "about to clone $(params.gitrepo.url) at $(params.gitrepo.commitish)"
+```
+
+In the above example, within a Task spec:
+
+```yaml
+- "-url=$(params.gitrepo.url)"
+```
+
+When populating a string field, It would be invalid (at least initially) to attempt to do variable replacement on the
+entire `gitrepo` object (`$(params.gitrepo)`). If we choose to support this later we could potentially replace the value
+with the json representation of the object directly.
+
+#### Variable replacement with object results
+
+#### Within a Task
+
+Within the context of a Task, Task authors would continue to have access to the `path` attribute of a result (identified
+by name) and would not have any additional variable replacement access to the keys.
+
+In the above example:
+
+```bash
+cat /place/with/actual/cloned/gitrepo.json > $(results.cloned-gitrepo.path)
```

Comment:
NIT: I think this is the only place where it's describe how object results will be produced by users - perhaps it would be worth having a small paragraph about it?

Reply:
there's also an example at https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md#emitting-object-results but happy to add more detail there and/or here (ill at least add a link from here back to that section) anything else in particular you'd like to see covered?